### PR TITLE
Support optional parameters at the end of the path

### PR DESCRIPTION
**Pull request by JulianBerger**
Originally wanted to add this feature to fastify: https://github.com/fastify/fastify/pull/2607 but we came to the conclusion that it fits better into here.

Support optional parameters (with a '?' suffix) if they are at the end of the path.

Example: 
**Input Route:** `/v1/user/:userId/admin/mail/:id?`
Can be accessed by:

1. `/v1/user/:userId/admin/mail/:id`
2. `/v1/user/:userId/admin/mail`

More information about the regex:
Screen Shot 2020-10-14 at 16 17 07

Screen Shot 2020-10-14 at 16 17 21
So the trailing slash is still preserved with the second capture group.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -70,6 +70,17 @@ Router.prototype.on = function on (method, path, opts, handler, store) {
   // handler validation
   assert(typeof handler === 'function', 'Handler should be a function')
 
+  // path ends with optional parameter
+  const optionalParamRegex = /(\/:[^/]*?)\?(\/?)$/
```
Comment (delvedor) [inline index.js:74]: Out of curiosity, why the need for a regex instead of `path.endsWith('?')`

Comment (JulianBerger) [reply index.js:74]: I had `path.endsWith('?') || path.endsWith('?/')` before, but then it's not ensured to be a parameter.
That would also match for routes like `/foo?`, don't know if that is an issue..

Comment (delvedor) [reply index.js:74]: Gotcha! 👍

**Review by delvedor — CHANGES_REQUESTED** [review]
Can you add some additional test cases?
Eg:
- what happens if someone declares two following routes?
  - `/foo/:bar?`
  - `/foo`
- what happens if someone puts the question mark inside a route declaration?
  - `/foo/:bar?/baz`

It's useful to test also "bad behaviors", so we can be sure not to break anything in the future.

**Review by mcollina — APPROVED** [review]
lgtm

File: test/optional-params.test.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+test('Test for param with ? not at the end', (t) => {
+  t.plan(1)
+  const findMyWay = FindMyWay({
+    defaultRoute: (req, res) => {
+      t.fail('Should not be defaultRoute')
+    }
+  })
+
+  findMyWay.on('GET', '/foo/:bar?/baz', (req, res, params) => {
```
Comment (delvedor) [inline test/optional-params.test.js:53]: Given that this case is not supported, I think we should throw an error here, what do you think?

File: index.js (hunk 1 above)
Comment (Eomm) [inline index.js:74]: moving this regexp near to `FULL_PATH_REGEXP` would avoid to recreate it for every route/`on` call. It should optimize a little bit the GC execution + remove the duplicated declaration*

File: index.js [hunk 3]
```diff
@@ -320,6 +331,17 @@ Router.prototype.off = function off (method, path) {
   assert(path.length > 0, 'The path could not be empty')
   assert(path[0] === '/' || path[0] === '*', 'The first character of a path should be `/` or `*`')
 
+  // path ends with optional parameter
+  const optionalParamRegex = /(\/:[^/]*?)\?(\/?)$/
```
Comment (Eomm) [inline index.js:335]: the duplicated declaration*

File: index.js (hunk 1 above)
Comment (mcollina) [reply index.js:74]: It would also avoid future bugs.

File: test/optional-params.test.js (hunk 2 above)
Comment (JulianBerger) [reply test/optional-params.test.js]: Yes, it's throwing an Error now.

File: index.js (hunk 1 above)
Comment (JulianBerger) [reply index.js]: Good idea, declaring it globally now.

**Review by JulianBerger — COMMENTED** [review]
All resolved

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by JulianBerger** [comment]
@delvedor can you merge it please? :)

**Review by delvedor — CHANGES_REQUESTED** [review]
Hello! Sorry for the delay :S
Everything looks good to me, can you add one last test to verify the behavior with `ignoreTrailingSlash: true`?

**Comment by JulianBerger** [comment]
Added tests for ignoreTrailingSlash.

File: test/optional-params.test.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+  findMyWay.on('GET', '/a/:p1-:p2?', (req, res, params) => {
+    if (params.p1 && params.p2) {
+      t.equal(params.p1, 'foo')
+      t.equal(params.p2, 'bar-baz')
+    }
+  })
+
+  findMyWay.lookup({ method: 'GET', url: '/a/foo-bar-baz', headers: {} }, null)
+  // findMyWay.lookup({ method: 'GET', url: '/a/foo', headers: {} }, null)
```
Comment (Eomm) [inline test/optional-params.test.js:77]:
```suggestion
```

File: test/optional-params.test.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  findMyWay.lookup({ method: 'GET', url: '/test/hello/foo', headers: {} }, null)
+  findMyWay.lookup({ method: 'GET', url: '/test/hello/foo/', headers: {} }, null)
+})
+
+test('Optional Parameter with ignoreTrailingSlash = false', (t) => {
+  t.plan(4)
+  const findMyWay = FindMyWay({
+    ignoreTrailingSlash: false,
+    defaultRoute: (req, res) => {
+      t.notMatch(req.url, '.*/$') // urls without trailing slash
```
Comment (Eomm) [inline test/optional-params.test.js:109]:
```suggestion
      t.fail('Should not be defaultRoute')
```

File: test/optional-params.test.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+test('Optional Parameter with ignoreTrailingSlash = true', (t) => {
+  t.plan(4)
+  const findMyWay = FindMyWay({
+    ignoreTrailingSlash: true,
+    defaultRoute: (req, res) => {
+      t.fail('Should not be defaultRoute')
+    }
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?/', (req, res, params) => {
```
Comment (delvedor) [inline test/optional-params.test.js:90]: There is no need to add the trailing slash in the route definition, it might be passed by the client, which you are testing correctly.

```suggestion
  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
```

File: test/optional-params.test.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+test('Optional Parameter with ignoreTrailingSlash = false', (t) => {
+  t.plan(4)
+  const findMyWay = FindMyWay({
+    ignoreTrailingSlash: false,
+    defaultRoute: (req, res) => {
+      t.notMatch(req.url, '.*/$') // urls without trailing slash
+    }
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?/', (req, res, params) => {
```
Comment (delvedor) [inline test/optional-params.test.js:113]:
```suggestion
  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
```

**Review by delvedor — CHANGES_REQUESTED** [review]
Great work, one little nit, you should not register the route with the trailing slash, it should be used only in the "request".

**Comment by JulianBerger** [comment]
> Great work, one little nit, you should not register the route with the trailing slash, it should be used only in the "request".

Ahh sure, fixed that! Also changed the ignoreTrailingSlash = false test behavior: `/test/hello/` will give us an empty string as the `:optional` parameter...

File: test/optional-params.test.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  t.plan(4)
+  const findMyWay = FindMyWay({
+    ignoreTrailingSlash: false,
+    defaultRoute: (req, res) => {
+      t.match(req.url, '/test/hello/foo/')
+    }
+  })
+
+  findMyWay.on('GET', '/test/hello/:optional?', (req, res, params) => {
+    console.log('req', req, 'opt', params.optional)
```
Comment (delvedor) [inline test/optional-params.test.js:113]:
```suggestion
```

File: test/optional-params.test.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    if (params.optional) {
```
Comment (delvedor) [inline test/optional-params.test.js:114]: This test looks brittle, can you write a more strict if check?

**Comment by delvedor** [comment]
@JulianBerger Hello! Do you have time to finish this pr? :)

**Comment by mcollina** [comment]
Closing for no activity

**Comment by jcward** [comment]
FYI for anyone who lands here looking for optional trailing parameters in find-my-way -- it suited my needs to simply register the same handler twice:

```
    findMyWay.on('GET', '/list/:id', handle_list_items);
    findMyWay.on('GET', '/list', handle_list_items);
```
